src/recursive.rs

```diff
@@ -128,24 +147,35 @@ impl<'a, 'b, I: Input<'a>, O, E: ParserExtra<'a, I>> Recursive<Indirect<'a, 'b,
             inner: RecursiveInner::Owned(RefC::new(Indirect {
                 inner: OnceCell::new(),
             })),
+            state: EmptyPhantom::new(),
         }
     }
+}
 
+impl<'a, 'b, I: Input<'a>, O, E: ParserExtra<'a, I>> Recursive<Indirect<'a, 'b, I, O, E>, Defined> {
     /// Defines the parser after declaring it, allowing it to be used for parsing.
     // INFO: Clone bound not actually needed, but good to be safe for future compat
     #[track_caller]
-    pub fn define<P: Parser<'a, I, O, E> + Clone + MaybeSync + 'a + 'b>(&mut self, parser: P) {
+    pub fn define<P: Parser<'a, I, O, E> + Clone + MaybeSync + 'a + 'b>(
+        parser: P,
+        declaration: Recursive<Indirect<'a, 'b, I, O, E>, Declared>,
+    ) -> Self {
```

Comment:
Should this function be on `Recursive<..., Declared>` and return a `Recursive<..., Defined>`?

Reply:
Yeah, you could do that too. For type inference it doesn’t matter, but it could make more sense logically.

Reply:
I'd say so, yes.

Reply:
Completed with 4572bb5